Userland/Libraries/LibVirtGPU/Device.cpp

```diff
@@ -1,21 +1,155 @@
+                                    "  1: MAD TEMP[1], IN[0].yyyy, CONST[1], TEMP[0]\n"
+                                    "  2: MAD TEMP[0], IN[0].zzzz, CONST[2], TEMP[1]\n"
+                                    "  3: MAD OUT[0], IN[0].wwww, CONST[3], TEMP[0]\n"
+                                    "  4: MOV_SAT OUT[1], IN[1]\n"
+                                    "  5: END\n"sv;
+
+Device::Device(NonnullRefPtr<Core::File> gpu_file, Gfx::IntSize size)
     : m_gpu_file { gpu_file }
 {
+    // Create a virgl context for this file descriptor
+    MUST(Core::System::ioctl(m_gpu_file->fd(), VIRGL_IOCTL_CREATE_CONTEXT));
+
+    // Create a VertexElements resource
+    VirGL3DResourceSpec vbo_spec {
+        .target = to_underlying(Gallium::PipeTextureTarget::BUFFER), // pipe_texture_target
+        .format = 0,                                                 // untyped buffer
+        .bind = to_underlying(Protocol::BindTarget::VIRGL_BIND_VERTEX_BUFFER),
+        .width = PAGE_SIZE * 256,
+        .height = 1,
+        .depth = 1,
+        .array_size = 1,
+        .last_level = 0,
+        .nr_samples = 0,
+        .flags = 0,
+        .created_resource_id = 0,
+    };
+    m_vbo_resource_id = create_virgl_resource(vbo_spec);
+
+    // Create a texture to draw to
+    VirGL3DResourceSpec drawtarget_spec {
+        .target = to_underlying(Gallium::PipeTextureTarget::TEXTURE_RECT),                  // pipe_texture_target
+        .format = to_underlying(Protocol::TextureFormat::VIRTIO_GPU_FORMAT_B8G8R8A8_UNORM), // pipe_to_virgl_format
+        .bind = to_underlying(Protocol::BindTarget::VIRGL_BIND_RENDER_TARGET),
+        .width = static_cast<u32>(size.width()),
+        .height = static_cast<u32>(size.height()),
+        .depth = 1,
+        .array_size = 1,
+        .last_level = 0,
+        .nr_samples = 0,
+        .flags = 0,
+        .created_resource_id = 0,
+    };
+    m_drawtarget = create_virgl_resource(drawtarget_spec);
+
+    // Create a depthbuffer surface
+    VirGL3DResourceSpec depthbuffer_surface_spec {
+        .target = to_underlying(Gallium::PipeTextureTarget::TEXTURE_RECT),             // pipe_texture_target
+        .format = to_underlying(Protocol::TextureFormat::VIRTIO_GPU_FORMAT_Z32_FLOAT), // pipe_to_virgl_format
+        .bind = to_underlying(Protocol::BindTarget::VIRGL_BIND_RENDER_TARGET) | to_underlying(Protocol::BindTarget::VIRGL_BIND_DEPTH_STENCIL),
+        .width = static_cast<u32>(size.width()),
+        .height = static_cast<u32>(size.height()),
+        .depth = 1,
+        .array_size = 1,
+        .last_level = 0,
+        .nr_samples = 0,
+        .flags = 0,
+        .created_resource_id = 0,
+    };
+    m_depthbuffer_surface = create_virgl_resource(depthbuffer_surface_spec);
+
+    // Initialize all required state
+    CommandBufferBuilder builder;
+
+    // Create and set the blend, to control the color mask
+    m_blend_handle = allocate_handle();
+    builder.append_create_blend(m_blend_handle);
```

Comment:
each of these appends (ENOMEM) + the command upload (ioctl errors) could all fail. probably worth a FIXME to make upload_command_buffer and create_virgl_resource return ErrorOr<void>, and to move this initialization to a factory or an initialize() method. 

MUST() from a constructor is not very nice.

Reply:
Context initialization is now in a separate method `initialize_context()`. The constructor only takes in the file now.